query-engine/core/src/query_graph_builder/write/utils.rs

```diff
@@ -1014,3 +1017,217 @@ pub fn emulate_on_update_cascade(
+                        .map(|(fk, pk)| {
+                            let fk_value = get_newest_fk_value(fk, &parent_fk_values, &fks_to_be_updated);
+
+                            pk.equals(fk_value)
+                        })
+                        .collect_vec();
+
+                    if let Node::Query(Query::Read(ReadQuery::ManyRecordsQuery(ref mut mrq))) =
+                        read_referenced_record_node
+                    {
+                        mrq.add_filter(Filter::and(filters));
+                    };
+
+                    Ok(read_referenced_record_node)
+                }),
+            ),
+        )?;
+
+        let ensure_referenced_record_exist_node = graph.create_node(Node::Empty);
+
+        graph.create_edge(
+            &read_referenced_record_node,
+            &ensure_referenced_record_exist_node,
+            QueryGraphDependency::ProjectedDataDependency(
+                child_model_identifier,
+                Box::new(|ensure_referenced_record_exist_node, referenced_records_ids| {
+                    // The record that would be referenced by Parent model _after_ the update is done does not exist.
+                    // This would break referential integrity, so we error out.
+                    if referenced_records_ids.is_empty() {
+                        return Err(QueryGraphBuilderError::RelationViolation((rf).into()));
+                    }
+
+                    Ok(ensure_referenced_record_exist_node)
+                }),
+            ),
+        )?;
+    }
+
+    Ok(())
+}
+
+/// Given a relation field (inline-side), find the foreign keys that are going to be updated (by value)
+fn find_fks_to_be_updated(
+    rf: RelationFieldRef,
+    update_args: &WriteArgs,
+) -> Option<(RelationFieldRef, HashMap<ScalarFieldRef, PrismaValue>)> {
+    let mut fks_to_be_updated: HashMap<ScalarFieldRef, PrismaValue> = HashMap::new();
+
+    for fk in rf.scalar_fields() {
+        if let Some(write_expr) = update_args.get_field_value(fk.db_name()) {
+            // If any of the update expressions are not values (but multiplications, divisions, etc..)
+            // Then skip the relation field entirely
+            if let Some(pv) = write_expr.as_value() {
+                fks_to_be_updated.insert(fk, pv.clone());
+            } else {
+                return None;
+            }
+        // If the foreign key is not part of the update args, then continue to the next one
+        } else {
+            continue;
+        }
+    }
+
+    // If no foreign keys were found in the update args, skip the relation field
+    if fks_to_be_updated.is_empty() {
+        return None;
+    }
+
+    Some((rf, fks_to_be_updated))
+}
+
+/// Given a set of old (existing foreign keys) and new foreign keys (ones that are about to be updated), pick the newest one
+fn get_newest_fk_value(
+    fk: ScalarFieldRef,
+    old_fks: &SelectionResult,
+    new_fks: &HashMap<ScalarFieldRef, PrismaValue>,
+) -> PrismaValue {
```

Comment:
The interface suggest this can't fail, but you're doing `unreachable!()` below. This is fine if you can explain why this will never be the case, but even then it's better interface design to return an option and to expect / unwrap it to signal that you're absolutely sure _at this point in the code_ that this can't fail.